README.md
# Sonar Quality Gate Maven Plugin

[![Maven Central](https://maven-badges.herokuapp.com/maven-central/io.github.r0bb3n/sonar-quality-gate-maven-plugin/badge.svg)](https://maven-badges.herokuapp.com/maven-central/io.github.r0bb3n/sonar-quality-gate-maven-plugin)
[![Quality Gate Status](https://sonarcloud.io/api/project_badges/measure?project=io.github.r0bb3n%3Asonar-quality-gate-maven-plugin&metric=alert_status)](https://sonarcloud.io/dashboard?id=io.github.r0bb3n%3Asonar-quality-gate-maven-plugin)
[![Coverage](https://sonarcloud.io/api/project_badges/measure?project=io.github.r0bb3n%3Asonar-quality-gate-maven-plugin&metric=coverage)](https://sonarcloud.io/dashboard?id=io.github.r0bb3n%3Asonar-quality-gate-maven-plugin)
[![Black Duck Security Risk](https://copilot.blackducksoftware.com/github/repos/r0bb3n/sonar-quality-gate-maven-plugin/branches/master/badge-risk.svg)](https://copilot.blackducksoftware.com/github/repos/r0bb3n/sonar-quality-gate-maven-plugin/branches/master)
![Maven Build](https://github.com/r0bb3n/sonar-quality-gate-maven-plugin/workflows/Maven%20Build/badge.svg)

Check your SonarQube project if it passes its quality gate. If it doesn't, the plugin will fail the maven job.

There a three modes supported:

| Mode                         | Description                                                                                                                                                                                          |
|------------------------------|------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| **integrated** (recommended) | Run in conjunction with sonar-maven-plugin (supports branches and pull requests)<br>The required information for fetching the data will be read from a sonar-maven-plugin generated file in `target` |
| **simple**                   | Run stand-alone for a simple SonarQube project (no branches)                                                                                                                                         |
| **advanced**                 | Run stand-alone for a SonarQube project with branch name or pull request                                                                                                                             |

## Prerequisites

* Java 11
  * see section [Why Java 11?](#why-java-11) for details
* SonarQube (depends on mode)
  * integrated: 5.3+ 
  * simple: 5.4+
  * advanced: 7.7+

## Usage

Include the plugin declaration in your `pom.xml`

```xml
<plugin>
  <groupId>io.github.r0bb3n</groupId>
  <artifactId>sonar-quality-gate-maven-plugin</artifactId>
  <version>1.3.0</version>
</plugin>
```

and execute the following command:

```
mvn sonar:sonar sonar-quality-gate:check
```

For **full reference** of the modes and parameters, please check the [Usage](https://r0bb3n.github.io/sonar-quality-gate-maven-plugin/usage.html)
and [Goals](https://r0bb3n.github.io/sonar-quality-gate-maven-plugin/plugin-info.html) page in the
[sonar-quality-gate-maven-plugin documentation](https://r0bb3n.github.io/sonar-quality-gate-maven-plugin/).

## Internals

### Why Java 11?

**TL;DR** usage of HttpClient (`java.net.http.HttpClient`)

Thinking about also supporting Java 8 led to the conclusion, that this would require relying on at least one additional
dependency (to maintain) for HTTP calls (e.g. Apache HttpClient), which is solved out-of-the-box in Java 11+ by simply
using `java.net.http.HttpClient`. Therefore, I decided to not support Java 8.

### Calling SonarQube Web API

#### Project Status

Get the quality gate status of a project.

**_Endpoint_**

`api/qualitygates/project_status`

**_URL Parameters_**
 
| Parameter     | Supported |
|---------------|-----------|
| `projectKey`  | yes       |
| `branch`      | yes       |
| `pullRequest` | yes       |
| `analysisId`  | yes       |
| `projectId`   | no        |

**_Documentation_** 
  
* on [sonarcloud.io Web-API](https://sonarcloud.io/web_api/api/qualitygates/project_status)
* in your SonarQube instance: `<sonar.host.url>/web_api/api/qualitygates/project_status`

#### Compute Engine Task (ceTask)

This endpoint is used to retrieve the analysis id of a prior sonar-maven-plugin run.

**_Endpoint_**

`api/ce/task`

**_URL Parameters_**
 
| Parameter          | Supported |
|--------------------|-----------|
| `id`               | yes       |
| `additionalFields` | no        |

**_Documentation_** 
  
* on [sonarcloud.io Web-API](https://sonarcloud.io/web_api/api/ce/task)
* in your SonarQube instance: `<sonar.host.url>/web_api/api/ce/task`

### Maven calls

#### Release 

**_prepare_** - manage `pom.xml` and create proper commits and tag and push to remote

```
mvn -B release:prepare -DdevelopmentVersion=1-SNAPSHOT -DreleaseVersion=<release version> -Dtag=v<release version>
```

**_perform_** - checkout version tag, create binaries and deploy via oss.sonatype.org to Maven Central and create
documentation ("site") and push to `gh-pages` branch.

```
mvn -B release:perform -DreleaseProfiles=build-for-release
```

**_clean_** - remove backup/work file (useful if you ran prepare but not perform)

```
mvn -B release:clean
```

### How-To release

0. Ensure the right git configs (username, email, signing stuff, ...) are set on `--global` level (correct `.gitconfig`
   file in user home) - config done on level `--local` is not respected by additional steps that take place in 
   subdirectories of `target` with fresh full-checkouts of the repo (e.g. commits/push during `site-deploy` via
   `maven-scm-publish-plugin`).
1. Update `CHANGELOG.md`: add a section for the upcoming version and move all "unpublished" changes to it
2. Update `README.md`: replace all occurrences of previous version number with upcoming version
3. Persist:
   `git add CHANGELOG.md README.md && git commit -m "prepare for release: update CHANGELOG.md/README.md" && git push`
4. Create release in git repo:
   `mvn -B release:prepare -DdevelopmentVersion=1-SNAPSHOT -DreleaseVersion=1.3.0 -Dtag=v1.3.0`
5. Create and publish binaries and documentation:
   `mvn -B release:perform -DreleaseProfiles=build-for-release`
6. Create and push a tag for the site branch. Change into directory `./target/checkout/target/scmpublish-checkout` and
   fire the according commands: `git tag v1.3.0-site && git push origin v1.3.0-site`
7. Create new release on GitHub ([here](https://github.com/r0bb3n/sonar-quality-gate-maven-plugin/releases/new))
   - choose tag: v1.3.0
   - set title: 1.3.0
   - copy the `CHANGELOG.md` content of the released version
   - upload the files `./target/checkout/target/*.(pom|jar|asc)`
8. publish [staging repository of oss nexus repository](https://oss.sonatype.org/#stagingRepositories)
   (login required) to maven
   central ([overview doc](https://central.sonatype.org/pages/ossrh-guide.html#releasing-to-central)
   / [detailed doc](https://central.sonatype.org/pages/releasing-the-deployment.html))
   1. check content of the staging repo
   2. select staging repo
   3. "Close" repo
   4. evaluation is now running, see repo tab "Activity"
   5. "Refresh" repo view
   6. "Release" repo
   7. copying is ongoing, see repo tab "Activity"
   8. "Refresh" repo view, staging repo will disappear after successful copy process
   9. verify that files are now available on the public release repo:
      [GAV search](https://oss.sonatype.org/#nexus-search;gav~io.github.r0bb3n~sonar-quality-gate-maven-plugin~1.*~~)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/), and this project adheres
to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

### Changed

- migrate to maven-plugin-report-plugin for site creation ([#379])

### Fixed

- fix broken "Fork me on GitHub" ribbon on plugin site ([#378])

[unreleased]: https://github.com/r0bb3n/sonar-quality-gate-maven-plugin/compare/v1.3.0...HEAD

[#378]: https://github.com/r0bb3n/sonar-quality-gate-maven-plugin/issues/378

[#379]: https://github.com/r0bb3n/sonar-quality-gate-maven-plugin/issues/379

## [1.3.0]

### Added

- support environment variable `SONAR_TOKEN` as alternative to the property `sonar.login` ([#243])

### Security

- update jackson-databind to fix [CVE-2022-42003] and [CVE-2022-42004] even if they are not exploitable in here

[1.3.0]: https://github.com/r0bb3n/sonar-quality-gate-maven-plugin/compare/v1.2.1...v1.3.0
[#243]: https://github.com/r0bb3n/sonar-quality-gate-maven-plugin/issues/243
[CVE-2022-42003]: https://nvd.nist.gov/vuln/detail/CVE-2022-42003
[CVE-2022-42004]: https://nvd.nist.gov/vuln/detail/CVE-2022-42004

## [1.2.1]

### Security

- update jackson-databind to fix [CVE-2020-36518] ([#203])

[1.2.1]: https://github.com/r0bb3n/sonar-quality-gate-maven-plugin/compare/v1.2.0...v1.2.1
[CVE-2020-36518]: https://nvd.nist.gov/vuln/detail/CVE-2020-36518
[#203]: https://github.com/r0bb3n/sonar-quality-gate-maven-plugin/issues/203

## [1.2.0]

### Added

- add parameter `skip` to support skipping the execution via config ([#115])
- add parameter `failOnMiss` to change failure behavior ([#118])
- publish maven plugin documentation to [r0bb3n.github.io/sonar-quality-gate-maven-plugin](https://r0bb3n.github.io/sonar-quality-gate-maven-plugin) ([#117])

[1.2.0]: https://github.com/r0bb3n/sonar-quality-gate-maven-plugin/compare/v1.1.0...v1.2.0
[#115]: https://github.com/r0bb3n/sonar-quality-gate-maven-plugin/issues/115
[#118]: https://github.com/r0bb3n/sonar-quality-gate-maven-plugin/issues/118
[#117]: https://github.com/r0bb3n/sonar-quality-gate-maven-plugin/issues/117

## [1.1.0]

### Added

- support multi module projects ([#55](https://github.com/r0bb3n/sonar-quality-gate-maven-plugin/issues/55))
- test coverage increased a lot ([#31](https://github.com/r0bb3n/sonar-quality-gate-maven-plugin/issues/31))

[1.1.0]: https://github.com/r0bb3n/sonar-quality-gate-maven-plugin/compare/v1.0.2...v1.1.0

## [1.0.2]

First version to be published to Maven Central. This requires some modification as
 described in the [Sonatype Guide][sonatype-guide].

### Changed

- `groupId` changed to `io.github.r0bb3n` to enable maven central deployment

[1.0.2]: https://github.com/r0bb3n/sonar-quality-gate-maven-plugin/compare/v1.0.1...v1.0.2

## [1.0.1]

First version to be published to Maven Central. This requires some modification as
 described in the [Sonatype Guide][sonatype-guide].

### Added

- provide `sources.jar` and `javadoc.jar` and signatures for all build artifacts automatically

### Changed

- `groupId` changed to `io.github.r0bb3n.maven` to enable maven central deployment

[1.0.1]: https://github.com/r0bb3n/sonar-quality-gate-maven-plugin/compare/v1.0.0...v1.0.1
[sonatype-guide]: https://central.sonatype.org/pages/producers.html

## [1.0.0]

### Added

- three modes defined: integrated, simple, advanced
- integrated mode: 
  - read metadata from previous run of sonar-maven-plugin to find correct analysis job
  - wait for analysis evaluated in SonarQube (polling for status updates)

### Changed

- parameters renamed to avoid unintended interferences with sonar-maven-plugin parameters

[1.0.0]: https://github.com/r0bb3n/sonar-quality-gate-maven-plugin/compare/v0.1.0...v1.0.0

## [0.1.0]

### Added

- check project status in SonarQube based on supplied parameters (e.g. branch name)
- authorization support for SonarQube

[0.1.0]: https://github.com/r0bb3n/sonar-quality-gate-maven-plugin/releases/tag/v0.1.0


src/main/java/org/r0bb3n/maven/SonarQualityGateMojo.java
/*
 * Copyright 2020 r0bb3n
 *
 * Licensed under the Apache License, Version 2.0 (the "License")
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.r0bb3n.maven;

import java.io.IOException;
import java.io.InputStream;
import java.net.URL;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Arrays;
import java.util.Optional;
import java.util.Properties;
import java.util.concurrent.TimeUnit;
import java.util.function.Predicate;
import java.util.stream.Collectors;
import lombok.SneakyThrows;
import org.apache.maven.plugin.AbstractMojo;
import org.apache.maven.plugin.MojoExecutionException;
import org.apache.maven.plugin.MojoFailureException;
import org.apache.maven.plugins.annotations.Mojo;
import org.apache.maven.plugins.annotations.Parameter;
import org.r0bb3n.maven.model.Condition;
import org.r0bb3n.maven.model.ProjectStatus;
import org.r0bb3n.maven.model.Task;

/**
 * Check project status in SonarQube and fail build, if quality gate is not passed
 * <br/>
 * <b>REQUIRED SonarQube version</b> (by mode):
 * <ul>
 *   <li>5.3+ (integrated)</li>
 *   <li>5.4+ (simple)</li>
 *   <li>7.7+ (advanced)</li>
 * </ul>
 */
@Mojo(name = "check", aggregator = true)
public class SonarQualityGateMojo extends AbstractMojo {

  private static final String PROP_SONAR_LOGIN = "sonar.login";
  private static final String PROP_SONAR_PASSWORD = "sonar.password";
  private static final String PROP_SONAR_HOST_URL = "sonar.host.url";
  private static final String ENV_SONAR_TOKEN = "SONAR_TOKEN";

  private static final String REPORT_TASK_KEY_CE_TASK_ID = "ceTaskId";

  /**
   * sonar host url<br/> (aligned to sonar-maven-plugin analysis parameters, see also
   * <a href="https://docs.sonarqube.org/latest/analysis/analysis-parameters/">SonarQube - Analysis
   * Parameters</a>)
   *
   * @since 1.0.0
   */
  @Parameter(property = PROP_SONAR_HOST_URL, defaultValue = "http://localhost:9000")
  private URL sonarHostUrl;

  /**
   * project key used in sonar for this project <br/>
   * <i>only used in modes: <b>simple</b>, <b>advanced</b></i><br/>
   * (aligned to sonar-maven-plugin analysis parameters, see also
   * <a href="https://docs.sonarqube.org/latest/analysis/analysis-parameters/">SonarQube - Analysis
   * Parameters</a>)
   *
   * @since 1.0.0
   */
  @Parameter(property = "sonar.projectKey",
      defaultValue = "${project.groupId}:${project.artifactId}")
  private String sonarProjectKey;

  /**
   * sonar login (username or token), see also
   * <a href="https://docs.sonarqube.org/latest/extend/web-api/">SonarQube
   * - Web API Authentication</a> <br/> (aligned to sonar-maven-plugin analysis parameters, see also
   * <a href="https://docs.sonarqube.org/latest/analysis/analysis-parameters/">SonarQube - Analysis
   * Parameters</a>)<br/>
   * Since version <b>1.3.0</b> instead the environment variable {@code SONAR_TOKEN} can be used to
   * provide a token (see these docs of
   * <a href="https://docs.sonarcloud.io/advanced-setup/ci-based-analysis/sonarscanner-for-maven/#configuration">SonarCloud</a>
   * and
   * <a href="https://docs.sonarqube.org/latest/analysis/github-integration/#header-3">SonarQube</a>
   * ).
   *
   * @since 1.0.0
   */
  @Parameter(property = PROP_SONAR_LOGIN)
  private String sonarLogin;

  /**
   * sonar password, see also <a href="https://docs.sonarqube.org/latest/extend/web-api/">SonarQube
   * - Web API Authentication</a> <br/> (aligned to sonar-maven-plugin analysis parameters, see also
   * <a href="https://docs.sonarqube.org/latest/analysis/analysis-parameters/">SonarQube - Analysis
   * Parameters</a>)
   *
   * @since 1.0.0
   */
  @Parameter(property = PROP_SONAR_PASSWORD)
  private String sonarPassword;

  /**
   * skip the execution of this plugin
   *
   * @since 1.2.0
   */
  @Parameter(property = "sonar-quality-gate.skip", defaultValue = "false")
  private boolean skip;

  /**
   * fail the execution, if the quality gate was not passed (not {@code OK})
   *
   * @since 1.2.0
   */
  @Parameter(property = "sonar-quality-gate.failOnMiss", defaultValue = "true")
  private boolean failOnMiss;

  /**
   * name of the branch to check the quality gate in sonar<br/>
   * <i>only used in mode: <b>advanced</b></i>
   *
   * @since 1.0.0
   */
  @Parameter(property = "sonar-quality-gate.branch")
  private String branch;

  /**
   * name of the pull request to check the quality gate in sonar<br/>
   * <i>only used in mode: <b>advanced</b></i>
   *
   * @since 1.0.0
   */
  @Parameter(property = "sonar-quality-gate.pullRequest")
  private String pullRequest;

  /**
   * How often try to retrieve the analysis id from the task details in sonar until stopping the
   * job<br/>
   * <i>only used in mode: <b>integrated</b></i>
   *
   * @since 1.0.0
   */
  @Parameter(property = "sonar-quality-gate.checkTask.attempts", defaultValue = "10")
  private int checkTaskAttempts;

  /**
   * How many seconds to wait between two requests when retrieving task details<br/>
   * <i>only used in mode: <b>integrated</b></i>
   *
   * @since 1.0.0
   */
  @Parameter(property = "sonar-quality-gate.checkTask.interval.s", defaultValue = "5")
  private int checkTaskIntervalS;

  /**
   * INTERNAL - get build directory
   */
  @Parameter(defaultValue = "${project.build.directory}", readonly = true)
  private String projectBuildDirectory;

  /**
   * connector instance to interact with sonar server
   */
  private SonarConnector sonarConnector;

  /**
   * request project status from sonar and evaluate quality gate result
   *
   * @throws MojoExecutionException configuration errors, io problems, ...
   * @throws MojoFailureException quality gate evaluates as not passed
   */
  public void execute() throws MojoExecutionException, MojoFailureException {
    if (skip) {
      getLog().info("skipped");
      return;
    }

    setupSonarConnector();

    String analysisId;
    if (Util.isBlank(branch) && Util.isBlank(pullRequest)) {
      Optional<String> ceTaskIdOpt = findCeTaskId(projectBuildDirectory);
      analysisId = ceTaskIdOpt
          // previous sonar run found, switching to 'integrated'
          .map(this::retrieveAnalysisId)
          // no previous sonar run found, switching to 'simple'
          .orElse(null);
    } else {
      // branch or PR was supplied, the 'advanced' mode was chosen
      analysisId = null;
    }

    ProjectStatus projectStatus = retrieveProjectStatus(analysisId);

    if (projectStatus.getStatus() != ProjectStatus.Status.OK) {
      String failedConditions = projectStatus.getConditions().stream()
          .filter(has(ProjectStatus.Status.OK, ProjectStatus.Status.NONE).negate())
          .map(c -> c.getMetricKey() + ":" + c.getStatus()).collect(Collectors.joining(", "));
      String message = String.format("Quality Gate not passed (status: %s)! Failed metric(s): %s",
          projectStatus.getStatus(), failedConditions);
      if (failOnMiss) {
        throw new MojoFailureException(message);
      } else {
        getLog().warn(message);
      }
    } else {
      getLog().info("project status: " + projectStatus.getStatus());
    }
  }

  /**
   * Read config parameters and create the {@link #sonarConnector}
   *
   * @throws MojoExecutionException in case of invalid config parameters
   */
  protected void setupSonarConnector() throws MojoExecutionException {
    if (!Util.isBlank(sonarPassword)) {
      // username+password auth
      if (Util.isBlank(sonarLogin)) {
        throw new MojoExecutionException(
            String.format("you cannot specify '%s' without '%s'", PROP_SONAR_PASSWORD,
                PROP_SONAR_LOGIN));
      } else {
        getLog().debug("sonar auth: username + password");
        sonarConnector =
            new SonarConnector(getLog(), sonarHostUrl, sonarProjectKey, sonarLogin, sonarPassword);
      }
    } else {
      // token auth
      if (!Util.isBlank(sonarLogin)) {
        getLog().debug("sonar auth: token (by property '" + PROP_SONAR_LOGIN + "')");
        sonarConnector =
            new SonarConnector(getLog(), sonarHostUrl, sonarProjectKey, sonarLogin, null);
      } else {
        // check environment variable SONAR_TOKEN as alternative source for the token
        String env = System.getenv(ENV_SONAR_TOKEN);
        if (!Util.isBlank(env)) {
          getLog().debug("sonar auth: token (by environment variable '" + ENV_SONAR_TOKEN + "')");
          sonarConnector = new SonarConnector(getLog(), sonarHostUrl, sonarProjectKey, env, null);
        }
      }
    }
    if (sonarConnector == null) {
      // no auth
      getLog().debug("sonar auth: none");
      sonarConnector = new SonarConnector(getLog(), sonarHostUrl, sonarProjectKey, null, null);
    }
  }

  /**
   * Call sonar server and retrieve the project status by either a recent analysis or by static
   * values for project, branch or pull request
   *
   * @param analysisId the actual analysis id to check for or {@code null} in case of
   *                   'simple' or 'advanced' mode
   * @return the project status
   * @throws MojoExecutionException in case of IO issues or interruption
   */
  protected ProjectStatus retrieveProjectStatus(String analysisId) throws MojoExecutionException {
    ProjectStatus projectStatus;
    try {
      if (analysisId != null) {
        // 'integrated' mode
        projectStatus = sonarConnector.retrieveProjectStatusByAnalysisId(analysisId);
      } else {
        // 'simple' / 'advanced' mode
        projectStatus = sonarConnector.retrieveProjectStatus(branch, pullRequest);
      }
    } catch (IOException e) {
      throw new MojoExecutionException("error fetching project status", e);
    } catch (InterruptedException e) {
      Thread.currentThread().interrupt();
      throw new MojoExecutionException("Interrupted while fetching project status", e);
    }
    return projectStatus;
  }

  /**
   * Check task details and read analysis id. If task is still ongoing ({@link
   * Task.Status#IN_PROGRESS}/{@link Task.Status#PENDING}), the threads sleeps for {@link
   * #checkTaskIntervalS} and tries in total {@link #checkTaskAttempts} times.
   *
   * <p>Throws MojoExecutionException when task got unsuitable status ({@link Task.Status#FAILED}/
   * {@link Task.Status#CANCELED}) or task is still ongoing but attempt limit is reached or IO
   * errors.
   *
   * @param ceTaskId ce task id to gather details (including analysis id)
   * @return analysis id, not null (unavailable id we cause an exception)
   */
  @SneakyThrows(MojoExecutionException.class)
  protected String retrieveAnalysisId(String ceTaskId) {
    int attemptsLeft = checkTaskAttempts;
    Task.Status status = Task.Status.IN_PROGRESS;
    String analysisId = null;

    while (status.isOngoing() && attemptsLeft-- > 0) {
      Task task;
      try {
        task = sonarConnector.retrieveTask(ceTaskId);
        status = task.getStatus();
      } catch (IOException e) {
        throw new MojoExecutionException("error while retrieving task", e);
      } catch (InterruptedException e) {
        Thread.currentThread().interrupt();
        throw new MojoExecutionException("Interrupted while retrieving task", e);
      }
      switch (status) {
        case SUCCESS:
          analysisId = task.getAnalysisId();
          break;
        case IN_PROGRESS:
        case PENDING:
          try {
            getLog().info(
                String.format("Analysis in progress, next retry in %ds (attempts left: %d)",
                    checkTaskIntervalS, attemptsLeft));
            Thread.sleep(TimeUnit.SECONDS.toMillis(checkTaskIntervalS));
          } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            throw new MojoExecutionException("Interrupted while waiting for retry", e);
          }
          break;
        default:
          throw new MojoExecutionException(
              "Cannot determine analysis id - unsuitable task status: " + status);
      }
    }
    if (analysisId == null) {
      throw new MojoExecutionException(String.format(
          "Could not fetch analysis id within %d requests with an interval of %d seconds (last "
              + "status: %s). Please increase the values 'checkTaskAttempts' and/or "
              + "'checkTaskIntervalS' to fit your projects needs.", checkTaskAttempts,
          checkTaskIntervalS, status));
    }

    return analysisId;
  }

  /**
   * Determine compute engine task id ("ceTaskId") of previous run of sonar-maven-plugin
   *
   * @param buildDir build directory, where sonar folder and report task file is located
   * @return id to request task details
   * @throws MojoExecutionException io problems when reading sonar-maven-plugin file
   */
  protected Optional<String> findCeTaskId(String buildDir) throws MojoExecutionException {
    Path reportTaskPath = Path.of(buildDir, "sonar", "report-task.txt");
    if (!Files.exists(reportTaskPath)) {
      getLog().info(
          "no report file from previously sonar-maven-plugin run found: " + reportTaskPath);
      return Optional.empty();
    }
    String ceTaskId;
    try (InputStream is = Files.newInputStream(reportTaskPath)) {
      Properties props = new Properties();
      props.load(is);
      ceTaskId = props.getProperty(REPORT_TASK_KEY_CE_TASK_ID);
    } catch (IOException e) {
      throw new MojoExecutionException(
          String.format("Error parsing properties in: %s", reportTaskPath), e);
    }
    if (Util.isBlank(ceTaskId)) {
      throw new MojoExecutionException(
          String.format("Property '%s' not found in '%s'", REPORT_TASK_KEY_CE_TASK_ID,
              reportTaskPath));
    } else {
      return Optional.of(ceTaskId);
    }
  }

  /**
   * create a predicate to check, if a {@link Condition} has one of the supplied status
   */
  private static Predicate<Condition> has(ProjectStatus.Status... status) {
    return c -> Arrays.asList(status).contains(c.getStatus());
  }

}
